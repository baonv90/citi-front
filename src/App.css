:root {
  --bg-color: #ffffff;
  --primary-color: #2D50E6;
  --title-color: #5d5d5d;
  --text-color: #7d7d7d;
  --skeleton-color-1: #d6d6d6;
  --skeleton-color-2: #cccccc;
  --skeleton-color-3: #e2e2e2;
  --grey-color: #f7f7f7;
  --light-color: #eeeeee;
  --inactive-color: #D95762;
  --active-color: #4CAF50;
  --border-color: #EAE9E3;
}

body {
  margin: 0;
  background-color: var(--bg-color);
  font-weight: 400;
  color: var(--text-color);
  font-size: 1rem;
  line-height: 1.5rem;
}