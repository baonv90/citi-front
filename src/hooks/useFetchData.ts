import { useState, useEffect } from 'react';
import { DataInterface } from '../interface';

// A hook to fetch data from given url
// it returns the fetching status loading and data
export const useFetchData = (url: RequestInfo) => {
  const [data, setData] = useState<null | DataInterface>();
  const [loading, setLoading] = useState<boolean>(true);
  const [hasError, setHasError] = useState<boolean>(false);

  useEffect(() => {
    fetch(url).then((res: Response) => {
      res.json().then((data) => {
        setData(data);
        setLoading(false);
      })
      .catch(() => {
        setHasError(true);
        setLoading(false);
      });
    });
  }, [url]);

  return { data, loading, hasError };
};
