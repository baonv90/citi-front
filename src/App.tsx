import React from 'react';
import styled from 'styled-components';

import Tabs from './layouts/Tabs';
import Insights from './components/Insights';
import DeviceList from './components/DeviceList';

import { getFilteredData } from './helpers/utils';
import { useFetchData } from './hooks/useFetchData';
import './App.css';
import { ResultProps } from './interface';

const URL = 'http://localhost:8010/devices';

const Container = styled.div`
  display: flex;
  padding: 1.5rem 5rem;
  flex-direction: column;
`;

const Title = styled.h1`
  font-size: 1.5rem;
  line-height: 2rem;
  font-weight: 600;
  color: var(--title-color);
`;

// tabs definition
const TABS = [
  { title: 'All Devices' },
  { title: 'Active', filter: 'connected' },
  { title: 'Inactive', filter: 'disconnected' }
];

function App () {
  const { data, loading, hasError } = useFetchData(URL);

  if (hasError) {
    return <div>Error while loading page...</div>;
  }

  let results: Array<ResultProps> = [];
  if (data) {
    results = data.results;
  }

  return (
    <Container>
      <Title>Insights</Title>
      <Insights isLoading={!data && loading} data={results}/>
      <Title>Devices</Title>
      <Tabs>
        {TABS.map(({ title, filter }, index) =>
          <DeviceList
            key={`tab-${index}`}
            title={title}
            data={getFilteredData(results, filter)}
            isLoading={!data && loading}
          />
        )}
      </Tabs>
    </Container>
  );
};

export default App;
