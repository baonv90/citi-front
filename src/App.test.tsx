import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';

import App from './App';
import { useFetchData } from './hooks/useFetchData';

jest.mock('./hooks/useFetchData');

describe('App', () => {
  const fakeData = {
    results: [
      { 
        status: 'connected',
        connection_type: 'wifi',
        last_seen_at: new Date().toISOString(),
        serial_number: 'Device_0'
      },
      { 
        status: 'disconnected',
        connection_type: 'celullar',
        last_seen_at: new Date().toISOString(),
        serial_number: 'Device_1'
      }
    ]
  }

  beforeEach(() => {
    useFetchData.mockReturnValue({
      loading: false,
      data: fakeData
    });
  });

  it('should match its snapshot', () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should contains all the elements', () => {
    render(<App />);
    expect(screen.getByText('Device_0')).toBeInTheDocument();
    expect(screen.getByText('wifi')).toBeInTheDocument();
    expect(screen.getByText('connected')).toBeInTheDocument();
  });

  it('should handle error while loading page', () => {
    useFetchData.mockReturnValue({
      loading: false,
      hasError: true
    });
    render(<App />);
    expect(screen.getByText('Error while loading page...')).toBeInTheDocument();
  });

  it('should handle tab navigation', () => {
    render(<App />);
    fireEvent.click(screen.getByText('Inactive'));
    expect(screen.getByText('Device_1')).toBeInTheDocument();
    expect(screen.queryByText('Device_0')).not.toBeInTheDocument();
  });
});
