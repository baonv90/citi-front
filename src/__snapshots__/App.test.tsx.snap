// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match its snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-dlMDgC cFbzzm"
  >
    <h1
      class="sc-kfYoZR PExsb"
    >
      Insights
    </h1>
    <div
      class="sc-fujyAs hQJiDC"
    >
      <div
        class="sc-eCApnc lazoQE"
      >
        <h3
          class="sc-jSFjdj gfrjyt"
        >
          Device Status
        </h3>
        <span
          class="sc-gKAaRy jQRRgX"
        >
          50% of all the devices are connected, 50% are disconnected
        </span>
      </div>
      <div
        class="sc-eCApnc lazoQE"
      >
        <h3
          class="sc-jSFjdj gfrjyt"
        >
          Connection
        </h3>
        <span
          class="sc-gKAaRy jQRRgX"
        >
          Of all the disconnected devices, celullar: 100% 
        </span>
      </div>
    </div>
    <h1
      class="sc-kfYoZR PExsb"
    >
      Devices
    </h1>
    <div
      class="sc-gtsrHT cPlHGU"
    >
      <div
        class="sc-dlnjwi GCjvm"
      >
        <div
          class="sc-bdnxRM ehFXTT"
        >
          All Devices
        </div>
        <div
          class="sc-bdnxRM eqRHDi"
        >
          Active
        </div>
        <div
          class="sc-bdnxRM eqRHDi"
        >
          Inactive
        </div>
      </div>
      <div
        class="sc-dIvrsQ jGygRc"
      >
        <div
          class="sc-crzoAE sc-fnVZcZ ga-DuxV gfoprS"
        >
          <div
            class="sc-hBMUJo sc-fFSPTT fDRiTP hZtpRi"
          >
            Serial Number
          </div>
          <div
            class="sc-hBMUJo sc-fFSPTT fDRiTP hZtpRi"
          >
            Status
          </div>
          <div
            class="sc-hBMUJo sc-fFSPTT fDRiTP hZtpRi"
          >
            Connection
          </div>
          <div
            class="sc-hBMUJo sc-fFSPTT fDRiTP hZtpRi"
          >
            Last update
          </div>
        </div>
        <div
          class="sc-hHEiqL jwqXKZ"
        >
          <div
            class="sc-iqAclL itlGMy"
          >
            <div
              class="sc-crzoAE ga-DuxV"
            >
              <div
                class="sc-hBMUJo fDRiTP"
              >
                Device_0
              </div>
              <div
                class="sc-pNWdM leQrRV"
              >
                <div
                  class="sc-jrsJWt cIQwkY"
                />
                <span
                  class="sc-kEqXSa jYkSXh"
                >
                  connected
                </span>
              </div>
              <div
                class="sc-hBMUJo fDRiTP"
              >
                wifi
              </div>
              <div
                class="sc-hBMUJo fDRiTP"
              >
                0 hour(s) ago
              </div>
            </div>
          </div>
          <div
            class="sc-iqAclL itlGMy"
          >
            <div
              class="sc-crzoAE ga-DuxV"
            >
              <div
                class="sc-hBMUJo fDRiTP"
              >
                Device_1
              </div>
              <div
                class="sc-pNWdM leQrRV"
              >
                <div
                  class="sc-jrsJWt daRvdX"
                />
                <span
                  class="sc-kEqXSa jYkSXh"
                >
                  disconnected
                </span>
              </div>
              <div
                class="sc-hBMUJo fDRiTP"
              >
                celullar
              </div>
              <div
                class="sc-hBMUJo fDRiTP"
              >
                0 hour(s) ago
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
